### PR TITLE
 TST: Test that `numpy.typing` can be imported in the absence of typing-extensions

### PR DESCRIPTION
**Pull request by BvB93**
xref https://github.com/numpy/numpy/issues/19521

On a semi-frequent we're running into issues with `numpy.typing` when the following 2 conditions are met:
* Python 3.7 is used.
* typing-extensions is not installed.

Ideally we'd just run a CI job without typing-extensions installed and see if the tests run as expected
but unfortunately this is not possible, as it is an (indirect) hard dependency of pytest. 

This PR instead opts for an alternative approach: the test introduced herein reloads the `numpy.typing` 
submodules after manually invalidating typing-extensions, thus making the package inaccessible as if it 
weren't installed in the first place.

### CODE REVIEW

File: numpy/typing/tests/test_typing_extensions.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    try:
+        sys.modules["typing_extensions"] = None
+        for name, module in MODULES.items():
+            _clear_attr(module)
+            assert importlib.reload(module), name
+    finally:
+        sys.modules["typing_extensions"] = typing_extensions
+        for module in MODULES.values():
+            _clear_attr(module)
+            importlib.reload(module)
```
Comment (BvB93) [inline numpy/typing/tests/test_typing_extensions.py:56]: Admitetly this is more of a hack then I'd ideally like it to be.
If anyone has suggestions for an alternative approach, then that would be very much welcome.

Comment (seberg) [reply numpy/typing/tests/test_typing_extensions.py:56]: Yeah, this is pretty terrifying :).  We have a `test_full_reimport` which does the same thing in a subprocess.  Can you do that instead?  If I understand it right, using `sys.modules["typing_extensions"] = None` is basically the "magic" ingredient here.  Moving it into a subprocess will remove the complexity of reimporting numpy proper.

Comment (BvB93) [reply numpy/typing/tests/test_typing_extensions.py:56]: >  If I understand it right, using `sys.modules["typing_extensions"] = None` is basically the "magic" ingredient here

Yes, by assigning an invalid value to the module an `ImportError` will be triggered whenever
executing `import typing_extensions` statements (which is the desired outcome here). 

I'm gona do a bit of experimenting with this `subprocess`-based approach, see if I can get it to work.

Comment (BvB93) [reply numpy/typing/tests/test_typing_extensions.py]: Alright, changed as of https://github.com/numpy/numpy/pull/19525/commits/ee9855823c41c0f5d4d2aa81e1965e898a375fbe.

I also pushed a small fix for `test_full_reimport` (https://github.com/numpy/numpy/pull/19525/commits/1ff314dc000cb0a7a1bb2823d3bc3ac4171f9d16). Namelly, exceptions raised within a subprocess 
stay within the subprocess, so they're effectively suppressed unless you actually validate the return code.

**Comment by seberg** [comment]
Looks fine to me and this is tests anyway, so putting it in.  Thanks Bas!

**Pull request merged**